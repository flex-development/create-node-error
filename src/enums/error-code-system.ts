/**
 * @file Enums - SystemErrorCode
 * @module errnode/enums/SystemErrorCode
 */

/**
 * System error codes.
 *
 * @see https://github.com/nodejs/node/blob/v19.3.0/deps/uv/include/uv.h#L72-L154
 *
 * @enum {Uppercase<string>}
 */
enum SystemErrorCode {
  E2BIG = 'E2BIG',
  EACCES = 'EACCES',
  EADDRINUSE = 'EADDRINUSE',
  EADDRNOTAVAIL = 'EADDRNOTAVAIL',
  EAFNOSUPPORT = 'EAFNOSUPPORT',
  EAGAIN = 'EAGAIN',
  EAI_ADDRFAMILY = 'EAI_ADDRFAMILY',
  EAI_AGAIN = 'EAI_AGAIN',
  EAI_BADFLAGS = 'EAI_BADFLAGS',
  EAI_BADHINTS = 'EAI_BADHINTS',
  EAI_CANCELED = 'EAI_CANCELED',
  EAI_FAIL = 'EAI_FAIL',
  EAI_FAMILY = 'EAI_FAMILY',
  EAI_MEMORY = 'EAI_MEMORY',
  EAI_NODATA = 'EAI_NODATA',
  EAI_NONAME = 'EAI_NONAME',
  EAI_OVERFLOW = 'EAI_OVERFLOW',
  EAI_PROTOCOL = 'EAI_PROTOCOL',
  EAI_SERVICE = 'EAI_SERVICE',
  EAI_SOCKTYPE = 'EAI_SOCKTYPE',
  EALREADY = 'EALREADY',
  EBADF = 'EBADF',
  EBADMSG = 'EBADMSG',
  EBUSY = 'EBUSY',
  ECANCELED = 'ECANCELED',
  ECHARSET = 'ECHARSET',
  ECHILD = 'ECHILD',
  ECONNABORTED = 'ECONNABORTED',
  ECONNREFUSED = 'ECONNREFUSED',
  ECONNRESET = 'ECONNRESET',
  EDEADLK = 'EDEADLK',
  EDESTADDRREQ = 'EDESTADDRREQ',
  EDOM = 'EDOM',
  EDQUOT = 'EDQUOT',
  EEXIST = 'EEXIST',
  EFAULT = 'EFAULT',
  EFBIG = 'EFBIG',
  EFTYPE = 'EFTYPE',
  EHOSTDOWN = 'EHOSTDOWN',
  EHOSTUNREACH = 'EHOSTUNREACH',
  EIDRM = 'EIDRM',
  EILSEQ = 'EILSEQ',
  EINPROGRESS = 'EINPROGRESS',
  EINTR = 'EINTR',
  EINVAL = 'EINVAL',
  EIO = 'EIO',
  EISCONN = 'EISCONN',
  EISDIR = 'EISDIR',
  ELOOP = 'ELOOP',
  EMFILE = 'EMFILE',
  EMLINK = 'EMLINK',
  EMSGSIZE = 'EMSGSIZE',
  EMULTIHOP = 'EMULTIHOP',
  ENAMETOOLONG = 'ENAMETOOLONG',
  ENETDOWN = 'ENETDOWN',
  ENETRESET = 'ENETRESET',
  ENETUNREACH = 'ENETUNREACH',
  ENFILE = 'ENFILE',
  ENOBUFS = 'ENOBUFS',
  ENODATA = 'ENODATA',
  ENODEV = 'ENODEV',
  ENOENT = 'ENOENT',
  ENOEXEC = 'ENOEXEC',
  ENOLCK = 'ENOLCK',
  ENOLINK = 'ENOLINK',
  ENOMEM = 'ENOMEM',
  ENOMSG = 'ENOMSG',
  ENONET = 'ENONET',
  ENOPROTOOPT = 'ENOPROTOOPT',
  ENOSPC = 'ENOSPC',
  ENOSR = 'ENOSR',
  ENOSTR = 'ENOSTR',
  ENOSYS = 'ENOSYS',
  ENOTCONN = 'ENOTCONN',
  ENOTDIR = 'ENOTDIR',
  ENOTEMPTY = 'ENOTEMPTY',
  ENOTSOCK = 'ENOTSOCK',
  ENOTSUP = 'ENOTSUP',
  ENOTTY = 'ENOTTY',
  ENXIO = 'ENXIO',
  EOF = 'EOF',
  EOPNOTSUPP = 'EOPNOTSUPP',
  EOVERFLOW = 'EOVERFLOW',
  EPERM = 'EPERM',
  EPIPE = 'EPIPE',
  EPROTO = 'EPROTO',
  EPROTONOSUPPORT = 'EPROTONOSUPPORT',
  EPROTOTYPE = 'EPROTOTYPE',
  ERANGE = 'ERANGE',
  EREMOTEIO = 'EREMOTEIO',
  EROFS = 'EROFS',
  ESHUTDOWN = 'ESHUTDOWN',
  ESOCKTNOSUPPORT = 'ESOCKTNOSUPPORT',
  ESPIPE = 'ESPIPE',
  ESRCH = 'ESRCH',
  ESTALE = 'ESTALE',
  ETIME = 'ETIME',
  ETIMEDOUT = 'ETIMEDOUT',
  ETXTBSY = 'ETXTBSY',
  EUNATCH = 'EUNATCH',
  EWOULDBLOCK = 'EWOULDBLOCK',
  EXDEV = 'EXDEV',
  UNKNOWN = 'UNKNOWN'
}

export default SystemErrorCode
