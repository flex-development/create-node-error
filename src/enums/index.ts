/**
 * @file Entry Point - Enums
 * @module errnode/enums
 */

export { default as codes } from './codes'
export { default as syscodes } from './syscodes'
