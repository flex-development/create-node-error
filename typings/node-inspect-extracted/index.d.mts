declare module 'node-inspect-extracted' {
  import type { ObjectPlain } from '@flex-development/tutils'

  export interface InspectOptions {
    breakLength?: number | undefined
    colors?: boolean | undefined
    compact?: boolean | number | undefined
    customInspect?: boolean | undefined
    depth?: number | null | undefined
    getters?: boolean | 'get' | 'set' | undefined
    maxArrayLength?: number | null | undefined
    maxStringLength?: number | null | undefined
    numericSeparator?: boolean | undefined
    showHidden?: boolean | undefined
    showProxy?: boolean | undefined
    sorted?: boolean | ((a: string, b: string) => number) | undefined
  }

  export interface InspectOptionsStylized extends InspectOptions {
    stylize(text: string, styleType: Style): string
  }

  export type Style =
    | 'bigint'
    | 'boolean'
    | 'date'
    | 'module'
    | 'null'
    | 'number'
    | 'regexp'
    | 'special'
    | 'string'
    | 'symbol'
    | 'undefined'

  export type CustomInspectFunction = (
    depth: number,
    options: InspectOptionsStylized
  ) => string

  export function format(format?: any, ...param: any[]): string

  export function formatWithOptions(
    inspectOptions: InspectOptions,
    format?: any,
    ...param: any[]
  ): string

  export function inspect(
    object: any,
    showHidden?: boolean,
    depth?: number | null,
    color?: boolean
  ): string

  export function inspect(object: any, options?: InspectOptions): string

  export namespace inspect {
    let colors: Record<number, number>
    let styles: { [K in Style]: string }
    let defaultOptions: InspectOptions
    let replDefaults: InspectOptions
    const custom: unique symbol
  }

  export function stripVTControlCharacters(str: string): string

  export function stylizeWithColor(text: string, styleType: Style): string

  export function stylizeWithHTML(text: string, styleType: Style): string

  export const Proxy: new <T extends ObjectPlain>(
    target: T,
    handler: ProxyHandler<T>
  ) => T
}
