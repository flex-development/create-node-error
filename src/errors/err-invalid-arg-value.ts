/**
 * @file Errors - ERR_INVALID_ARG_VALUE
 * @module errnode/errors/ERR_INVALID_ARG_VALUE
 * @see https://github.com/nodejs/node/blob/v22.7.0/lib/internal/errors.js#L1444-L1451
 */

import E from '#e'
import { codes } from '#src/enums'
import type { NodeError, NodeErrorConstructor } from '#src/interfaces'
import { DOT, truncate } from '@flex-development/tutils'
import { inspect } from 'node-inspect-extracted'

/**
 * `ERR_INVALID_ARG_VALUE` schema.
 *
 * @see {@linkcode NodeError}
 * @see https://nodejs.org/api/errors.html#err_invalid_arg_value
 *
 * @extends {NodeError<codes.ERR_INVALID_ARG_VALUE>}
 * @extends {TypeError}
 */
interface ErrInvalidArgValue
  extends NodeError<codes.ERR_INVALID_ARG_VALUE>, TypeError {}

/**
 * `ERR_INVALID_ARG_VALUE` message arguments.
 */
type Args = [name: string, value: unknown, reason?: string | null | undefined]

/**
 * `ERR_INVALID_ARG_VALUE` constructor.
 *
 * @see {@linkcode Args}
 * @see {@linkcode ErrInvalidArgValue}
 * @see {@linkcode NodeErrorConstructor}
 *
 * @extends {NodeErrorConstructor<ErrInvalidArgValue,Args>}
 */
interface ErrInvalidArgValueConstructor
  extends NodeErrorConstructor<ErrInvalidArgValue, Args> {
  /**
   * Create a new `ERR_INVALID_ARG_VALUE` error.
   *
   * @see {@linkcode ErrInvalidArgValue}
   *
   * @param {string} name
   *  Name of invalid argument or property
   * @param {unknown} value
   *  The invalid value
   * @param {string | null | undefined} [reason]
   *  Reason for invalidity
   * @return {ErrInvalidArgValue}
   */
  new (
    name: string,
    value: unknown,
    reason?: string | null | undefined
  ): ErrInvalidArgValue
}

/**
 * `ERR_INVALID_ARG_VALUE` model.
 *
 * Thrown when an invalid or unsupported value is passed for a given argument or
 * property.
 *
 * @see {@linkcode ErrInvalidArgValueConstructor}
 *
 * @type {ErrInvalidArgValueConstructor}
 *
 * @class
 */
const ERR_INVALID_ARG_VALUE: ErrInvalidArgValueConstructor = E(
  codes.ERR_INVALID_ARG_VALUE,
  TypeError,
  /**
   * @param {string} name
   *  Name of invalid argument or property
   * @param {unknown} value
   *  The invalid value
   * @param {string | null | undefined} [reason]
   *  Reason for invalidity
   * @return {string}
   *  Error message
   */
  function message(
    name: string,
    value: unknown,
    reason: string | null | undefined = null
  ): string {
    /**
     * Inspected {@linkcode value}.
     *
     * @var {string} inspected
     */
    let inspected: string = inspect(value)

    /**
     * Error message.
     *
     * @var {string} message
     */
    let message: string = 'The'

    message += ` ${name.includes(DOT) ? 'property' : 'argument'} '${name}'`

    if ((reason ??= 'is invalid')) {
      message += ` ${reason}`
      message = message.trim()
    }

    /* c8 ignore next 3 */ if (inspected.length > 128) {
      inspected = truncate(inspected, 128) + DOT.repeat(3)
    }

    message += `${DOT} Received ${inspected}`

    return message
  }
)

export {
  ERR_INVALID_ARG_VALUE as default,
  type ErrInvalidArgValue,
  type Args as ErrInvalidArgValueArgs,
  type ErrInvalidArgValueConstructor
}
