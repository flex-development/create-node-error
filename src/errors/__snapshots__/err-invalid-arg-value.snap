// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:errors/ERR_INVALID_ARG_VALUE > should build message from parameters (0) 1`] = `"The argument 'hints' is invalid. Received 3"`;

exports[`unit:errors/ERR_INVALID_ARG_VALUE > should build message from parameters (1) 1`] = `"The property 'options.objectMode' is not supported. Received true"`;
